src/stacked_borrows/stack.rs

```diff
@@ -0,0 +1,372 @@
+        };
+
+        if let Some(idx) = self.find_granting_tagged(access, tag) {
+            return Ok(Some(idx));
+        }
+
+        // Couldn't find it in the stack; but if there is an unknown bottom it might be there.
+        let found = self.unknown_bottom.is_some_and(|&unknown_limit| {
+            tag.0 < unknown_limit.0 // unknown_limit is an upper bound for what can be in the unknown bottom.
+        });
+        if found { Ok(None) } else { Err(()) }
+    }
+
+    fn find_granting_tagged(&mut self, access: AccessKind, tag: SbTag) -> Option<usize> {
+        #[cfg(feature = "stack-cache")]
+        if let Some(idx) = self.find_granting_cache(access, tag) {
+            return Some(idx);
+        }
+
+        // If we didn't find the tag in the cache, fall back to a linear search of the
+        // whole stack, and add the tag to the stack.
+        for (stack_idx, item) in self.borrows.iter().enumerate().rev() {
+            if tag == item.tag && item.perm.grants(access) {
+                #[cfg(feature = "stack-cache")]
+                self.cache.add(stack_idx, tag);
+                return Some(stack_idx);
+            }
+        }
+        None
+    }
+
+    #[cfg(feature = "stack-cache")]
+    fn find_granting_cache(&mut self, access: AccessKind, tag: SbTag) -> Option<usize> {
+        // This looks like a common-sense optimization; we're going to do a linear search of the
+        // cache or the borrow stack to scan the shorter of the two. This optimization is miniscule
+        // and this check actually ensures we do not access an invalid cache.
+        // When a stack is created and when tags are removed from the top of the borrow stack, we
+        // need some valid value to populate the cache. In both cases, we try to use the bottom
+        // item. But when the stack is cleared in `set_unknown_bottom` there is nothing we could
+        // place in the cache that is correct. But due to the way we populate the cache in
+        // `StackCache::add`, we know that when the borrow stack has grown larger than the cache,
+        // every slot in the cache is valid.
+        if self.borrows.len() <= CACHE_LEN {
+            return None;
+        }
+        // Search the cache for the tag we're looking up
+        let cache_idx = self.cache.tags.iter().position(|t| *t == tag)?;
+        let stack_idx = self.cache.idx[cache_idx];
+        // If we found the tag, look up its position in the stack to see if it grants
+        // the required permission
+        if self.borrows[stack_idx].perm.grants(access) {
+            // If it does, and it's not already in the most-recently-used position, move it there.
+            // Except if the tag is in position 1, this is equivalent to just a swap, so do that.
+            if cache_idx == 1 {
+                self.cache.tags.swap(0, 1);
+                self.cache.idx.swap(0, 1);
+            } else if cache_idx > 1 {
+                self.cache.add(stack_idx, tag);
```

Comment:
This doesn't "move" it to the top, it adds it to the top and shifts everything, so now it is in the cache twice, right?

Reply:
Oh goodness, you're right. I threw together the better implementation, and it is a bit surprising to me that this doesn't seem to matter. It's possible that because I've already special-cased lookups of the first 2 elements in the cache, which should be by far the most commonly used, this duplication just doesn't happen very much? So at the moment, I just explained in a comment what's going on here as best I can, and I favor keeping the implementation simple.

Reply:
Sounds good.